fix(testing): async/fakeAsync/inject/withModule helpers should pass through context to callback functions

BendingBender opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
When using async/fakeAsync/inject/withModule test helpers, context (`this`) is not passed through to callback fn. This breaks the usage of Jasmine's variable sharing system.

**What is the new behavior?**
Make sure that context (`this`) that is passed to functions generated by test helpers is passed through to the callback functions. Enables usage of Jasmine's variable sharing system to prevent accidental memory leaks.

**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**:

vicb reviewed (approved):
Could you please add a quick comment on each modified arrow function, something like:
`// Not using an arrow function to preserve the context`

BendingBender commented on modules/@angular/core/testing/test_bed.ts:
```diff
@@ -412,17 +412,17 @@ export function getTestBed() {
+      // the injected tokens.
+      return testBed.compileComponents().then(() => {
+        const completer: AsyncTestCompleter = testBed.get(AsyncTestCompleter);
+        testBed.execute(tokens, fn.bind(this));
```
Instead of doing `fn.bind(this)` I would have preferred to change the call signature of `testBed.execute()` from
```
execute(tokens: any[], fn: Function): any
```
to
```
execute(tokens: any[], fn: Function, context?: any): any
```

But I believe this would constitute a breaking change on the public API. What do you think, is it desireable or not?

BendingBender committed:
fix(testing): async/fakeAsync/inject/withModule helpers should pass through context to callback functions

 Make sure that context (`this`) that is passed to functions generated by test helpers is passed through to the callback functions. Enables usage of Jasmine's variable sharing system to prevent accidental memory leaks during test runs.

BendingBender commented on modules/@angular/core/testing/test_bed.ts:
```diff
@@ -412,17 +412,19 @@ export function getTestBed() {
+      // the injected tokens.
+      return testBed.compileComponents().then(() => {
+        const completer: AsyncTestCompleter = testBed.get(AsyncTestCompleter);
+        testBed.execute(tokens, fn.bind(this));
```
Instead of doing `fn.bind(this)` I would have preferred to change the call signature of `testBed.execute()` from
```
execute(tokens: any[], fn: Function): any
```
to
```
execute(tokens: any[], fn: Function, context?: any): any
```

But I believe this would constitute a breaking change on the public API. What do you think, is it desirable or not?

vicb replied:
The 2nd option looks better to me.

While we can not commit breaking change to the 2.x branch, we can in the 4.x branch.

What you can do is:
- Keep the first commit as is (with the label `fix(): ...`),
- Add a second `feat():...` commit that will be merged only in the 4.x branch

BendingBender reviewed (commented):
Fixed comments around modified arrow functions.

BendingBender committed:
feat(testing): add optional `context` param to `TestBed.execute()` and refactor call sites

BendingBender commented on modules/@angular/core/testing/test_bed.ts:
```diff
@@ -324,10 +324,10 @@ export class TestBed implements Injector {
   }
 
-  execute(tokens: any[], fn: Function): any {
+  execute(tokens: any[], fn: Function, context?: any): any {
```
Changed as suggested. Tested it out, this is not breaking but needs a minor version bump.

vicb commented:
@juliemr @vikerman could one of you please double check and apply the merge label - changes LGTM. Thanks.

mhevery commented:
merged in: 5f40e5ba21b2311dbd904a396b7a4cb11a92c179
